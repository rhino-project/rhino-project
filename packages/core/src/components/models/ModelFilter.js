import { useModel } from '../../hooks/models';
import {
  getModelAndAttributeFromPath,
  getModelFromRef,
  isOwnerGlobal
} from '../../utils/models';
import { ModelFilterBoolean } from './filters/ModelFilterBoolean';
import { ModelFilterDate } from './filters/ModelFilterDate';
import { ModelFilterDateTime } from './filters/ModelFilterDateTime';
import { ModelFilterEnum } from './filters/ModelFilterEnum';
import { ModelFilterFloat } from './filters/ModelFilterFloat';
import { ModelFilterInteger } from './filters/ModelFilterInteger';
import { ModelFilterIntegerSelect } from './filters/ModelFilterIntegerSelect';
import { ModelFilterOwnerReference } from './filters/ModelFilterOwnerReference';
import { ModelFilterReference } from './filters/ModelFilterReference';
import { ModelFilterString } from './filters/ModelFilterString';
import { ModelFilterTime } from './filters/ModelFilterTime';
import { ModelFilterYear } from './filters/ModelFilterYear';
import { useGlobalComponentForModel } from '../../hooks/overrides';

export const ModelFilterBase = (props) => {
  const model = useModel(props.model);
  const { path } = props;

  const [, attribute] = getModelAndAttributeFromPath(model, path);

  // FIXME: Make this a separate function so that its easier to override
  switch (attribute?.type) {
    case 'boolean':
      return <ModelFilterBoolean {...props} />;
    case 'integer':
      switch (attribute.format) {
        case 'year':
          return <ModelFilterYear {...props} />;
        case 'select':
          return <ModelFilterIntegerSelect {...props} />;
        default:
          return <ModelFilterInteger {...props} />;
      }
    case 'float':
      return <ModelFilterFloat {...props} />;

    case 'reference':
      // eslint-disable-next-line no-case-declarations
      const refModel = getModelFromRef(attribute);

      if (isOwnerGlobal(refModel)) return <ModelFilterReference {...props} />;

      return <ModelFilterOwnerReference {...props} />;
    case 'string':
      if (attribute?.enum) return <ModelFilterEnum {...props} />;

      switch (attribute.format) {
        case 'date':
          return <ModelFilterDate {...props} />;
        case 'time':
          return <ModelFilterTime {...props} />;
        case 'datetime':
          return <ModelFilterDateTime {...props} />;
        default:
          return <ModelFilterString {...props} />;
      }
    default:
      console.assert(false, 'No available filter for ', attribute);
  }

  return 'No filter for this attribute type';
};

export const ModelFilter = (props) =>
  useGlobalComponentForModel('ModelFilter', ModelFilterBase, props);
