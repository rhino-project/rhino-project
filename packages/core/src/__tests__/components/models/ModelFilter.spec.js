import ModelFilter from 'rhino/components/models/ModelFilter';
import { sharedModelTests } from './sharedModelTests';

/* eslint react/display-name: 0, react/prop-types: 0 */
describe('ModelFilter', () => {
  sharedModelTests(ModelFilter);
});
