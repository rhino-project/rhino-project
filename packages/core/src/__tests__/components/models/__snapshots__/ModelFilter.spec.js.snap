// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelFilter should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelFilter should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
